oak/server/rust/oak_runtime/src/node/wasm.rs

```diff
@@ -0,0 +1,785 @@
+                }
+                ChannelEither::Writer(channel) => {
+                    self.writers.insert(handle, channel);
+                }
+            }
+
+            return handle;
+        }
+    }
+
+    /// Helper function to get memory.
+    fn get_memory(&self) -> &wasmi::MemoryRef {
+        self.memory
+            .as_ref()
+            .expect("WasmInterface memory not attached!?")
+    }
+
+    /// Make sure a given address falls within the currently allocated range of memory.
+    fn validate_ptr(&self, addr: AbiPointer, offset: AbiPointerOffset) -> Result<(), OakStatus> {
+        let byte_size: wasmi::memory_units::Bytes = self.get_memory().current_size().into();
+
+        if byte_size < wasmi::memory_units::Bytes((addr as usize) + (offset as usize)) {
+            return Err(OakStatus::ERR_INVALID_ARGS);
+        }
+
+        Ok(())
+    }
+
+    /// Creates a new `WasmInterface` structure.
+    pub fn new(
+        pretty_name: String,
+        runtime: RuntimeRef,
+        initial_handle: ChannelReader,
+    ) -> WasmInterface {
+        WasmInterface {
+            pretty_name,
+            runtime,
+            readers: {
+                let mut r = HashMap::new();
+                r.insert(1, initial_handle);
+                r
+            },
+            writers: HashMap::new(),
+            memory: None,
+        }
+    }
+
+    /// Corresponds to the host ABI function `node_create: (usize, usize, usize, usize, u64) ->
+    /// u32`.
+    fn node_create(
+        &self,
+        name_ptr: AbiPointer,
+        name_length: AbiPointerOffset,
+        entrypoint_ptr: AbiPointer,
+        entrypoint_length: AbiPointerOffset,
+        initial_handle: AbiHandle,
+    ) -> Result<(), OakStatus> {
+        let config_name = self
+            .get_memory()
+            .get(name_ptr, name_length as usize)
+            .map_err(|err| {
+                error!(
+                    "node_create: Unable to read name from guest memory: {:?}",
+                    err
+                );
+                OakStatus::ERR_INVALID_ARGS
+            })?;
+
+        let config_name = String::from_utf8(config_name).map_err(|err| {
+            error!("node_create: Unable to parse config_name: {:?}", err);
+            OakStatus::ERR_INVALID_ARGS
+        })?;
+
+        let entrypoint = self
+            .get_memory()
+            .get(entrypoint_ptr, entrypoint_length as usize)
+            .map_err(|err| {
+                error!(
+                    "node_create: Unable to read entrypoint from guest memory: {:?}",
+                    err
+                );
+                OakStatus::ERR_INVALID_ARGS
+            })?;
+
+        let entrypoint = String::from_utf8(entrypoint).map_err(|err| {
+            error!("node_create: Unable to parse entrypoint: {:?}", err);
+            OakStatus::ERR_INVALID_ARGS
+        })?;
+
+        let handle = self.readers.get(&initial_handle).ok_or(()).map_err(|_| {
```

Comment:
nit: this isn't a `handle` now we're in the runtime; maybe `channel_ref` or `reader` or somesuch instead?